optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Gets the deep cloning decision for a field with a particular value assigned to it.
+     *
+     * @param solutionDescriptor The solution descriptor for the solution class
+     * @param field The field to get the deep cloning decision of
+     * @param owningClass The class that owns the field; can be different
+     *        from the field's declaring class (ex: subclass).
+     * @param actualValueClass The class of the value that is currently assigned
+     *        to the field; can be different from the field type
+     *        (ex: for the field "List myList", the actual value
+     *        class might be ArrayList).
+     * @return True iff the field should be deep cloned with a particular value.
+     */
+    public static boolean getDeepCloneDecision(SolutionDescriptor<?> solutionDescriptor,
+            Field field, Class<?> owningClass,
+            Class<?> actualValueClass) {
+        return isFieldDeepCloned(solutionDescriptor, field, owningClass) || isClassDeepCloned(solutionDescriptor,
+                actualValueClass);
+    }
+
+    /**
+     * Gets the deep cloning decision for a field.
+     *
+     * @param solutionDescriptor The solution descriptor for the solution class
+     * @param field The field to get the deep cloning decision of
+     * @param owningClass The class that owns the field; can be different
+     *        from the field's declaring class (ex: subclass).
+     * @return True iff the field should always be deep cloned (regardless of value).
+     */
+    public static boolean isFieldDeepCloned(SolutionDescriptor<?> solutionDescriptor,
+            Field field, Class<?> owningClass) {
+        if (field.getType().isEnum()) {
+            return false;
+        }
+        return isFieldAnEntityPropertyOnSolution(solutionDescriptor, field, owningClass)
+                || isFieldAnEntityOrSolution(solutionDescriptor, field)
+                || isFieldADeepCloneProperty(field, owningClass);
+    }
+
+    /**
+     * Return true only if a field represent an entity property on the solution class.
+     * An entity property is one who type is a PlanningEntity or a collection
+     * of PlanningEntity.
+     *
+     * @param solutionDescriptor The solution descriptor for the solution class
+     * @param field The field to get the deep cloning decision of
+     * @param owningClass The class that owns the field; can be different
+     *        from the field's declaring class (ex: subclass).
+     * @return True only if the field is an entity property on the solution class.
+     *         May return false if the field getter/setter is complex.
+     */
+    public static boolean isFieldAnEntityPropertyOnSolution(SolutionDescriptor<?> solutionDescriptor,
+            Field field, Class<?> owningClass) {
+        // field.getDeclaringClass() is a superclass of or equal to the owningClass
+        if (solutionDescriptor.getSolutionClass().isAssignableFrom(owningClass)) {
```

Comment:
Personally, I'd invert the logic here.
Then the most of this method would not need to be nested, making it easier to reason about.